File: examples/classes.js
#!/usr/bin/env node
/**
 * JavaScript Classes and Object-Oriented Programming Example
 * Demonstrates classes, inheritance, and methods
 */

// Simple class
class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  distanceFromOrigin() {
    return Math.sqrt(this.x ** 2 + this.y ** 2);
  }

  toString() {
    return `Point(${this.x}, ${this.y})`;
  }
}

// Class with inheritance
class Shape {
  constructor(name) {
    this.name = name;
  }

  describe() {
    return `This is a ${this.name}`;
  }
}

class Rectangle extends Shape {
  constructor(width, height) {
    super("rectangle");
    this.width = width;
    this.height = height;
  }

  area() {
    return this.width * this.height;
  }

  perimeter() {
    return 2 * (this.width + this.height);
  }

  toJSON() {
    return {
      name: this.name,
      width: this.width,
      height: this.height,
      area: this.area(),
      perimeter: this.perimeter()
    };
  }
}

// Main execution
function main() {
  // Test Point
  const p1 = new Point(3, 4);
  console.log(`Point: ${p1}`);
  console.log(`Distance from origin: ${p1.distanceFromOrigin().toFixed(2)}`);

  // Test Rectangle
  const rect = new Rectangle(10, 5);
  console.log(`\n${rect.describe()}`);
  console.log(`Width: ${rect.width}, Height: ${rect.height}`);
  console.log(`Area: ${rect.area()}`);
  console.log(`Perimeter: ${rect.perimeter()}`);

  console.log(`\nRectangle as JSON:`);
  console.log(JSON.stringify(rect.toJSON(), null, 2));

  // Array of points
  const points = Array.from({ length: 3 }, (_, i) => new Point(i + 1, (i + 1) * 2));
  console.log(`\nPoints: [${points.join(", ")}]`);
  console.log(`Distances: [${points.map(p => p.distanceFromOrigin().toFixed(2)).join(", ")}]`);
}

main();

File: examples/classes.py
#!/usr/bin/env python3
"""
Python Classes and Object-Oriented Programming Example
Demonstrates classes, inheritance, and methods
"""

from dataclasses import dataclass
from typing import List

# Using dataclass for simpler class definition
@dataclass
class Point:
    x: float
    y: float

    def distance_from_origin(self):
        return (self.x ** 2 + self.y ** 2) ** 0.5

    def __str__(self):
        return f"Point({self.x}, {self.y})"


# Traditional class with inheritance
class Shape:
    def __init__(self, name):
        self.name = name

    def describe(self):
        return f"This is a {self.name}"


class Rectangle(Shape):
    def __init__(self, width, height):
        super().__init__("rectangle")
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

    def perimeter(self):
        return 2 * (self.width + self.height)


# Main execution
if __name__ == "__main__":
    # Test Point
    p1 = Point(3, 4)
    print(f"Point: {p1}")
    print(f"Distance from origin: {p1.distance_from_origin():.2f}")

    # Test Rectangle
    rect = Rectangle(10, 5)
    print(f"\n{rect.describe()}")
    print(f"Width: {rect.width}, Height: {rect.height}")
    print(f"Area: {rect.area()}")
    print(f"Perimeter: {rect.perimeter()}")

    # List comprehension with objects
    points = [Point(i, i*2) for i in range(1, 4)]
    print(f"\nPoints: {[str(p) for p in points]}")
    print(f"Distances: {[p.distance_from_origin() for p in points]}")
